### DISCUSSION TITLE
How can I detect the default locale based on the URL when using localePrefix = 'as-needed'

### DISCUSSION THREAD

**Discussion by staffanmowitz** [amannn/next-intl#876] [Q&A]
I am working on a project that involves two languages: Swedish and English. I am using prefix-based routing with ﻿`localePrefix` set to ﻿`as-needed`, where Swedish is the default language. When manually switching the language to English and accessing a Swedish URL, _/test_, it redirects to _/en/test._

As per the documentation, the locale detection process works like this:

> 1. A locale prefix is present in the pathname (e.g. /en/about)
> 2. A cookie is present that contains a previously detected locale
> 3. A locale can be matched based on the [accept-language header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language)
> 4. As a last resort, the defaultLocale is used

Since my setup lacks a locale prefix in the URL pathname due to ﻿`localePrefix` being set to `﻿as-needed`, next-intl will move on to the next step and detect the locale from the cookie.

Is there a way to make next-intl recognize URLs without a locale prefix as Swedish and change the locale accordingly?

**Comment by amannn** [comment]
Sure: [`localeDetection: false`](https://next-intl-docs.vercel.app/docs/routing/middleware#locale-detection)

**Reply by staffanmowitz** [reply]
Thanks for the quick reply!

I still want to keep the ability to detect the locale from headers and cookies.

I'm basically looking for a way to detect that the locale is Swedish when there is no locale prefix in the URL.

Is this possible?

**Reply by amannn** [reply]
Hmm, I'm unsure I can follow how the locale negotiation should work in this case. Can you provide some examples of what you expect to happen for requests for a few pathnames, potentially with a cookie from a previous visit and considering locale prefixes? Do you want to turn off redirects that consider a previous locale for all pathnames but the root of the app (`/`)?

But to answer the question, you can compose the middleware and read a potential locale prefix from the pathname, similar to how it's shown in the [additional rewrites example](https://next-intl-docs.vercel.app/docs/routing/middleware#example-additional-rewrites). Based on that, you modify the request (e.g. remove/set a cookie) or use that information to configure the `next-intl` middleware on a case-by-case base.

Does that help?

**Reply by staffanmowitz** [reply]
The problem occurs when a user who has previously visited the site, and has the cookie set to `en`, tries to visit a page with Swedish locale, i.e. a URL that has no locale prefix in it.

For example, I have a Swedish news article with the URL *example.com/**nyheter/nyhetsartikel***. If the user visits this URL but has the cookie set to `en`, they will end up on the URL *example.com/**en/news/nyhetsartikel***, which does not exist. If the user does *not* have the cookie set to `en` it works as expected.

I have read the documentation on how to [configure the middleware matcher to detect pathnames without locale prefixes](https://next-intl-docs.vercel.app/docs/routing/middleware#matcher-no-prefix) and applied the suggested matcher pattern, but I still can't make it work.
